executor/executor.go

```diff
@@ -5,175 +5,228 @@ package executor
 
 import (
 	"sync"
+	"sync/atomic"
 
 	"github.com/ava-labs/avalanchego/utils/set"
 
 	"github.com/ava-labs/hypersdk/state"
+
+	uatomic "go.uber.org/atomic"
 )
 
-const defaultSetSize = 8
+// baseDependencies must be greater than the maximum number of dependencies
+// any single task could have. This is used to ensure a dependent task
+// does not begin executing a task until all dependencies have been enqueued.
+const baseDependencies = 100_000_000
 
 // Executor sequences the concurrent execution of
 // tasks with arbitrary conflicts on-the-fly.
 //
 // Executor ensures that conflicting tasks
 // are executed in the order they were queued.
 // Tasks with no conflicts are executed immediately.
+//
+// It is assumed that no single task has more than [baseDependencies]. If
+// this invariant is violated, some tasks will never execute and this code
+// could deadlock.
 type Executor struct {
-	metrics    Metrics
-	wg         sync.WaitGroup
-	executable chan *task
-
-	stop     chan struct{}
-	err      error
-	stopOnce sync.Once
-
-	l         sync.Mutex
-	done      bool
-	completed int
-	tasks     map[int]*task
-	edges     map[string]int
+	metrics Metrics
+
+	workers sync.WaitGroup
+
+	outstanding sync.WaitGroup
+	executable  chan *task
+
+	tasks int
+	nodes map[string]*task
+
+	err uatomic.Error
 }
 
 // New creates a new [Executor].
 func New(items, concurrency int, metrics Metrics) *Executor {
 	e := &Executor{
 		metrics:    metrics,
-		stop:       make(chan struct{}),
-		tasks:      make(map[int]*task, items),
-		edges:      make(map[string]int, items*2), // TODO: tune this
-		executable: make(chan *task, items),       // ensure we don't block while holding lock
+		nodes:      make(map[string]*task, items*2), // TODO: tune this
+		executable: make(chan *task, items),         // ensure we don't block while holding lock
 	}
+	e.workers.Add(concurrency)
 	for i := 0; i < concurrency; i++ {
-		e.createWorker()
+		go e.work()
 	}
 	return e
 }
 
+func (e *Executor) work() {
+	defer e.workers.Done()
+
+	for {
+		t, ok := <-e.executable
+		if !ok {
+			return
+		}
+		e.runTask(t)
+	}
+}
+
 type task struct {
 	id int
 	f  func() error
+	// reading are the tasks that this task is using non-exclusively.
+	reading []*task
 
-	dependencies set.Set[int]
-	blocking     set.Set[int]
-
+	l        sync.Mutex
 	executed bool
+	// blocked are the tasks that are waiting on this task to execute
+	// before they can be executed. tasks should not be added to
+	// [blocked] after the task has been executed.
+	blocked map[int]*task
```

Comment:
maybe `exclusive`? I don't really like `exclusive` replacing `blocked` but I figured I'd suggest it, as it then makes the exclusive <> shared key relationship explicit and clear

Reply:
I don't think `exclusive` makes sense for this because non-exclusive requests are also stored here.